### ISSUE TITLE
filter2 command is too slow

### ISSUE THREAD

**Issue by y9c** [shenwei356/csvtk#269]
Compared with filter command or awk, fiter2 command is much slower, especially for rule with multiple conditions. 

It might be relative to this function in the for-loop, which repeatedly parsing the expression.
https://github.com/shenwei356/csvtk/blob/9407f73e2d72dddf5042c7dbb6299a180ea9cf4a/csvtk/cmd/filter2.go#L370-L376

**Comment by shenwei356** [comment]
Yes, I noticed that. It is slow :(

**Comment by y9c** [comment]
Can we move the Expression parsing function outside the for-loop and run it only once?

**Comment by shenwei356** [comment]
It must be done. Cause `filterStr1` is different in each iteration.

**Comment by y9c** [comment]
Why filterStr1 is different? Can we cache the parsed results?

**Comment by shenwei356** [comment]
It's the expression, like '$age > 18', the `$age` needs to be replaced with the value of each row.

**Comment by y9c** [comment]
Yes. I mean can we parsed the expression as something like '$1>18' and reuse the code of the filter command to deal with the computation afterward

**Comment by shenwei356** [comment]
> parsed the expression as something like '$1>18' and reuse the code of the filter command 
I don't think so.

God, it's really slow~ I used it a lot recently. Have to improve it, when I have time ~

**Comment by shenwei356** [comment]
It's fixed. Please update it to v0.31.1. https://github.com/shenwei356/csvtk/releases/tag/v0.31.1
